src/student_success_tool/modeling/bias_detection.py

```diff
@@ -0,0 +1,213 @@
+    "🔴 HIGH BIAS": "high_bias",
+}
+
+
+def calculate_fnpr_and_ci(
+    targets: pd.Series,
+    preds: pd.Series,
+    min_fnpr_samples: int = MIN_FNPR_SAMPLES,
+) -> tuple[float, float, float]:
+    """
+    Calculates the False Negative Prediction Rate (FNPR) and its confidence interval.
+
+    Args:
+        targets: Labels from model output
+        preds: Predictions from model output
+        min_fnpr_samples: Minimum number of true positives or false negatives for FNPR calculation. When TP or FN are low, FNPR can be very unstable and flag subgroups based on small differences in TP or FN.
+    Returns:
+        Tuple of FNPR, lower CI bound, and upper CI bound.
+    """
+    cm = confusion_matrix(targets, preds, labels=[False, True])
+    tn, fp, fn, tp = (
+        cm.ravel()
+        if cm.shape == (2, 2)
+        else np.bincount(np.array(targets) * 2 + np.array(preds), minlength=4)
+    )
+
+    denominator = fn + tp
+
+    if (tp < min_fnpr_samples) or (fn < min_fnpr_samples):
+        return np.nan, np.nan, np.nan
+
+    fnpr = fn / denominator
+    margin = Z * np.sqrt((fnpr * (1 - fnpr)) / denominator)
+    return fnpr, max(0, fnpr - margin), min(1, fnpr + margin)
+
+
+def check_ci_overlap(
+    ci1: tuple[float, float],
+    ci2: tuple[float, float],
+) -> bool:
+    """
+    Checks whether confidence intervals (CIs) overlap. If they do, the FNPR differences
+    are within the margin of error at the 95% confidence level. If the CIs do not
+    overlap, this suggests strong statistical evidence that the FNPRs are different.
+
+    Args:
+        ci1: Confidence interval (min, max) for subgroup 1
+        ci2: Confidence interval (min, max) for subgroup 2
+    Returns:
+        Boolean indicating whether the CIs overlap.
+    """
+    return not (ci1[1] < ci2[0] or ci2[1] < ci1[0])
+
+
+def z_test_fnpr_difference(
+    fnpr1: float,
+    fnpr2: float,
+    denominator1: int,
+    denominator2: int,
+) -> float:
+    """
+    Performs a z-test for the FNPR difference between two groups. If there are
+    less than 30 samples of false negatives and true negatives, then we do not
+    have enough data to perform a z-test. Thirty samples is the standard check
+    for z-tests.
+
+    Args:
+        fnpr1: FNPR value for subgroup 1
+        fnpr2: FNPR value for subgroup 2
+        denominator1: Number of false negatives + true negatives for subgroup 1
+        denominator2: Number of false negatives + true negatives for subgroup 2
+    Returns:
+        Two-tailed p-value for the z-test for the FNPR difference between the two subgroups.
+    """
+    if (
+        denominator1 <= 30 or denominator2 <= 30
+    ):  # Ensures valid sample sizes for z-test
+        return np.nan
+    std_error = np.sqrt(
+        (fnpr1 * (1 - fnpr1)) / denominator1 + (fnpr2 * (1 - fnpr2)) / denominator2
+    )
+    z_stat = (fnpr1 - fnpr2) / std_error
+    return float(2 * (1 - st.norm.cdf(abs(z_stat))))  # Two-tailed p-value
```

Comment:
I _think_ this calculation is preferred:

```suggestion
    return 2.0 * st.norm.sf(abs(z_stat))  # Two-tailed p-value
```